packages/com.github.actions.contrib/ActionGenerator.pkl

```diff
@@ -113,34 +112,36 @@ local actionYaml =
 local effectiveAction =
   action ?? deepToTyped.apply(Action.getClass(), yamlParser.parse(actionYaml)) as Action
 
-local actionTypesUrl1 =
-  if (subpath == null)
-    "https://raw.githubusercontent.com/\(owner)/\(repo)/refs/tags/\(version)/action-types.yml"
+local versionMajor = semver.parseOrNull(version.replaceFirst("v", ""))?.major
+local actionTypesUrlFallbackVersion = versionMajor.ifNonNull((it) -> "v\(versionMajor)")
+
+local actionTypesUrls: Listing<String?> = new {
+  "https://raw.githubusercontent.com/\(owner)/\(repo)/refs/tags/\(version)\(subpath.ifNonNull((it) -> "/\(it)") ?? "")/action-types.yml"
+  if (actionTypesUrlFallbackVersion != null)
+    "https://raw.githubusercontent.com/\(owner)/\(repo)/refs/tags/\(actionTypesUrlFallbackVersion)\(subpath.ifNonNull((it) -> "/\(it)") ?? "")/action-types.yml"
   else
-    "https://raw.githubusercontent.com/\(owner)/\(repo)/refs/tags/\(version)/\(subpath)/action-types.yml"
+    null
 
-local actionTypesUrl2 =
-  if (subpath == null)
-    "https://raw.githubusercontent.com/typesafegithub/github-actions-typing-catalog/refs/heads/main/typings/\(owner)/\(repo)/\(version)/action-types.yml"
+  "https://raw.githubusercontent.com/typesafegithub/github-actions-typing-catalog/refs/heads/main/typings/\(owner)/\(repo)/\(version)\(subpath.ifNonNull((it) -> "/\(it)") ?? "")/action-types.yml"
+  if (actionTypesUrlFallbackVersion != null)
+    "https://raw.githubusercontent.com/typesafegithub/github-actions-typing-catalog/refs/heads/main/typings/\(owner)/\(repo)/\(actionTypesUrlFallbackVersion)\(subpath.ifNonNull((it) -> "/\(it)") ?? "")/action-types.yml"
   else
-    "https://raw.githubusercontent.com/typesafegithub/github-actions-typing-catalog/refs/heads/main/typings/\(owner)/\(repo)/\(version)/\(subpath)/action-types.yml"
+    null
+}
+
+local actionTypesUrl =
+  actionTypesUrls.toList().filterNonNull().findOrNull((it) -> read?(it) != null)
```

Comment:
Do we need a [`firstNotNullOf`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/first-not-null-of.html)? 👀 
```suggestion
  actionTypesUrls.toList().filterNonNull().firstNotNullOf((it) -> read?(it))
```